### PR TITLE
Allow user to provide HttpClient instance in OData Client

### PR DESCRIPTION
**Pull request by habbes**
### Issues

*This pull request fixes #xxx.*

### Description

*Briefly describe the changes of this pull request.*

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Client/DataServiceContext.cs [hunk 1]
```diff
@@ -253,6 +274,25 @@ public DataServiceContext(Uri serviceRoot, ODataProtocolVersion maxProtocolVersi
...
+        /// A Uri provided with a trailing slash is equivalent to one without such a trailing character.
+        /// With Silverlight, the <paramref name="serviceRoot"/> can be a relative Uri
+        /// that will be combined with System.Windows.Browser.HtmlPage.Document.DocumentUri.
+        /// </remarks>
+        public DataServiceContext(Uri serviceRoot, ODataProtocolVersion maxProtocolVersion, IHttpClientHandlerProvider httpClientHandlerProvider)
+            : this(serviceRoot, maxProtocolVersion, ClientEdmModelCache.GetModel(maxProtocolVersion), httpClientHandlerProvider: httpClientHandlerProvider)
+        {
+        }
+
         /// <summary>
```
Comment (mikepizzo) [inline src/Microsoft.OData.Client/DataServiceContext.cs:296]: Given that the httpclienthandlerprovider is settable on the DataServiceContext, is there any reason we need to add constructor overloads for it?

File: src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+{
+    /// <summary>
+    /// Provides the <see cref="HttpClientHandler"/> instance(s) used by OData Client
+    /// to execute requests.
+    /// </summary>
+    public interface IHttpClientHandlerProvider
+    {
+        /// <summary>
+        /// Returns an <see cref="HttpClientHandler"/> instance that OData Client should
+        /// used to eecute a request. This method is called per request.
```
Comment (mikepizzo) [inline src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs:19]: >used to eecute [](http://example.com/codeflow?start=12&length=14)

"used to eecute"=>"used to execute"

File: src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        /// </summary>
+        /// <returns>The <see cref="HttpClientHandler"/> instanced that will be used to execute the request.</returns>
```
Comment (mikepizzo) [inline src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs:21]: >instanced [](http://example.com/codeflow?start=57&length=9)

"instanced"=>"instance"

File: src/Microsoft.OData.Client/DataServiceClientRequestMessageArgs.cs [hunk 4]
```diff
@@ -81,5 +96,10 @@ public string ActualMethod
         /// Gets a System.Boolean value that controls whether default credentials are sent with requests.
         /// </summary>
         public bool UseDefaultCredentials { get; private set; }
+
+        /// <summary>
+        /// Returns the <see cref="HttpClientHandler"/> that should be used to send the request message.
```
Comment (mikepizzo) [inline src/Microsoft.OData.Client/DataServiceClientRequestMessageArgs.cs:101]: >Returns [](http://example.com/codeflow?start=12&length=7)

Maybe change "Returns" to "Gets" for consistency.

**Comment by mikepizzo** [comment]
A few minor comments, but otherwise looks good to me. Main question is whether we need to add constructor overloads for DataServiceContext, or can just call the setter for the IHttpClientHandlerProvider.

**Review by mikepizzo — CHANGES_REQUESTED** [review]
:clock1:

File: src/Microsoft.OData.Client/DataServiceContext.cs (hunk 1 above)
Comment (habbes) [reply src/Microsoft.OData.Client/DataServiceContext.cs:296]: Not it's not necessary. Just that I started with the constructor argument then added the setter after the fact.

Since the `HttpRequestTransportMode` is set via property and not constructor, it does make sense to keep this on the property only.

File: src/Microsoft.OData.Client/DataServiceClientRequestMessageArgs.cs [hunk 5]
```diff
@@ -28,6 +28,20 @@ public class DataServiceClientRequestMessageArgs
...
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DataServiceClientRequestMessageArgs"/> class.
+        /// </summary>
+        /// <param name="method">Method of the request.</param>
+        /// <param name="requestUri">The Request Uri.</param>
+        /// <param name="useDefaultCredentials">True if the default credentials need to be sent with the request. Otherwise false.</param>
+        /// <param name="usePostTunneling">True if the request message must use POST verb for the request and pass the actual verb in X-HTTP-Method header, otherwise false.</param>
+        /// <param name="headers">The set of headers for the request.</param>
+        /// <param name="httpClientHandlerProvider">Returns the <see cref="HttpClientHandler"/> that should be used to send the request message.</param>
```
[the remark this answers is not in the document]
Comment (habbes) [reply src/Microsoft.OData.Client/DataServiceClientRequestMessageArgs.cs:43]: I meant to say: The httpClientHandlerProvider returns the httpClientHandler...

Let me rephrase it.

**Review by mikepizzo — APPROVED** [review]
:shipit:

File: src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs [hunk 6]
```diff
@@ -50,6 +52,7 @@ public class HttpClientRequestMessage : DataServiceClientRequestMessage, ISendin
         private readonly HttpRequestMessage _requestMessage;
         private readonly HttpClient _client;
         private readonly HttpClientHandler _handler;
+        private readonly IHttpClientHandlerProvider _clientHandlerProvider;
```
Comment (corranrogue9) [inline src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs:55]:
```suggestion
```

I don't think we need a member for this, it seems we only use it in the constructor

File: src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs [hunk 7]
```diff
@@ -64,19 +67,31 @@ public class HttpClientRequestMessage : DataServiceClientRequestMessage, ISendin
...
+            if (_clientHandlerProvider == null)
+            {
+                _handler = new HttpClientHandler();
+                _client = new HttpClient(_handler, disposeHandler: true);
+            }
+            else
+            {
+                _handler = _clientHandlerProvider.GetHttpClientHandler();
+                _client = new HttpClient(_handler, disposeHandler: false);
+            }
```
Comment (corranrogue9) [inline src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs:93]:
```suggestion
            var clientHandlerProvider = args.HttpClientHandlerProvider;
            if (clientHandlerProvider == null)
            {
                this._handler = new HttpClientHandler();
                this._client = new HttpClient(this._handler, disposeHandler: true);
            }
            else
            {
                try
                {
                    this._handler = clientHandlerProvider.GetHttpClientHandler();
                    this._client = new HttpClient(this._handler, disposeHandler: false);
                }
                catch
                {
                    this._messageStream.Dispose();
                    throw;
                }
            }
```

The `HttpClient` constructor will throw in case of a `null` handler, plus we likely don't own the provider code, so it could throw anything really.

File: src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs [hunk 8]
```diff
@@ -514,11 +529,23 @@ public void Dispose()
...
+
+            _disposed = true;
+
+            if (disposing)
+            {
+                HttpResponseMessage response = _httpResponseMessage;
+                _httpResponseMessage = null;
+                if (response != null)
+                {
+                    ((IDisposable)response).Dispose();
```
Comment (corranrogue9) [inline src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs:545]: I have a feeling the cast here gets JITed away anyway, but I'm curious where this comes from. It's in the original code too. I've only seem something like this with an explicit interface implementation or something. Do you know why we are doing this cast?

File: src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs [hunk 9]
```diff
@@ -514,11 +529,23 @@ public void Dispose()
         /// <param name="disposing">If 'true' this method is called from user code; if 'false' it is called by the runtime.</param>
         protected virtual void Dispose(bool disposing)
         {
-            HttpResponseMessage response = _httpResponseMessage;
-            _httpResponseMessage = null;
-            if (response != null)
+            if (_disposed)
```
Comment (corranrogue9) [inline src/Microsoft.OData.Client/Serialization/HttpClientRequestMessage.cs:532]:
```suggestion
            if (this._disposed)
```

File: src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        /// <summary>
+        /// Returns an <see cref="HttpClientHandler"/> instance that OData Client should
+        /// used to execute a request. This method is called per request.
+        /// </summary>
+        /// <returns>The <see cref="HttpClientHandler"/> instance that will be used to execute the request.</returns>
+        /// <remarks>
+        /// It's possible that the returned <see cref="HttpClientHandler"/> may be mutated inside a request hook
+        /// e.g. via <see cref="HttpClientRequestMessage.Credentials"/>.
+        /// OData Client does not dispose the <see cref="HttpClientHandler"/> instance.
```
Comment (corranrogue9) [inline src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs:25]: The lack of disposing is the entire point of the providing, right? It seems like this should be noted in the interface summary and the method summary rather than a remark, since it's critical to both the performance improvement and the memory management that the implementer understands this.

File: src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        /// </remarks>
+        HttpClientHandler GetHttpClientHandler();
```
Comment (corranrogue9) [inline src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs:27]: Did we want to provide the context or args for this? It seems like it will be difficult for the implementer to know when to re-use a `HttpClientHandler` instance or when to provide a new one. (this is truly a discussion question, I think I don't know enough about `HttpClientHandler` and `HttpClient` to know the answer).

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs [hunk 12]
```diff
@@ -18,5 +28,185 @@ public void UnwrapAggregateException()
...
+                    usePostTunneling: false,
+                    new Dictionary<string, string>(),
+                    httpClientHandlerProvider);
+
+                using (var request = new HttpClientRequestMessage(args))
+                {
+                    // Act
+                    IODataResponseMessage response = request.GetResponse();
+                    var stream = response.GetStream();
+                    string contents = "";
```
Comment (corranrogue9) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs:55]:
```suggestion
                    string contents = "";
                    using (var stream = response.GetStream())
```

Better to be explicit about this stuff I think...

Comment (corranrogue9) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs:55]:
```suggestion
                    string contents = "";
                    using (IODataResponseMessage response = request.GetResponse())
                    using (var stream = response.GetStream())
```

Better to be explicit about this stuff I think...

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs [hunk 13]
```diff
@@ -18,5 +28,185 @@ public void UnwrapAggregateException()
...
+                var newResponse = await client.GetAsync("http://foo");
+
+                // Assert
+                // We can still make request from the client after HttpClientRequestMessage is disposed
+                var contents = await newResponse.Content.ReadAsStringAsync();
+                Assert.Equal(expectedResponse, contents);
+                Assert.Equal(2, handler.Requests.Count);
+                Assert.Equal("GET http://localhost/", handler.Requests[0]);
+                Assert.Equal("GET http://foo/", handler.Requests[1]);
+                Assert.False(handler.Disposed);
```
Comment (corranrogue9) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs:107]:
```suggestion
                using (var request = new HttpClientRequestMessage(args))
                {
                    // Act
                    IODataResponseMessage response = request.GetResponse();
                    var stream = response.GetStream();
                    stream.Dispose();
                    request.Dispose();

                    // Re-use handler with a different HttpClient after initial request is disposed
                    using (var client = new HttpClient(handler, disposeHandler: false))
                    using (var newResponse = await client.GetAsync("http://foo"))
                    {
                        // Assert
                        // We can still make request from the client after HttpClientRequestMessage is disposed
                        var contents = await newResponse.Content.ReadAsStringAsync();
                        Assert.Equal(expectedResponse, contents);
                        Assert.Equal(2, handler.Requests.Count);
                        Assert.Equal("GET http://localhost/", handler.Requests[0]);
                        Assert.Equal("GET http://foo/", handler.Requests[1]);
                        Assert.False(handler.Disposed);
                    }
                }
```

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockHttpClientHandler.cs [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        /// </summary>
+        /// <param name="expectedResponse">The contents to return in each response.</param>
+        public MockHttpClientHandler(string expectedResponse)
+        {
+            _requestHandler = (_) =>
+            {
+                var response = new HttpResponseMessage(HttpStatusCode.OK);
+                response.Content = new StringContent(expectedResponse);
+                return response;
+            };
```
Comment (corranrogue9) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockHttpClientHandler.cs:39]:
```suggestion
        : this((_) =>
            {
                var response = new HttpResponseMessage(HttpStatusCode.OK);
                response.Content = new StringContent(expectedResponse);
                return response;
            })
        {
```

I'm not sure if this actually compiles, or that it's even readable, but I think there's something to be said about the explicit nature of having a primary constructor.

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockHttpClientHandlerProvider.cs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    /// </summary>
+    internal sealed class MockHttpClientHandlerProvider: IHttpClientHandlerProvider
+    {
+        private readonly HttpClientHandler _handler;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="MockHttpClientHandlerProvider"/>.
+        /// </summary>
+        /// <param name="client">The <see cref="HttpClientHandler"/> to return.</param>
+        public MockHttpClientHandlerProvider(HttpClientHandler handler, bool requestClientDisposal = false)
```
Comment (corranrogue9) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockHttpClientHandlerProvider.cs:23]:
```suggestion
        public MockHttpClientHandlerProvider(HttpClientHandler handler)
```

I think that was probably just left over from something else? I'm not seeing it used anywhere

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockHttpClientHandlerProvider.cs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+        /// <summary>
+        /// Number of times <see cref="GetHttpClientHandler()"/> was
+        /// called.
+        /// </summary>
+        public int NumCalls { get; private set; } = 0;
+
+        public HttpClientHandler GetHttpClientHandler()
+        {
+            NumCalls++;
```
Comment (corranrogue9) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockHttpClientHandlerProvider.cs:36]: I suppose we aren't concerned about thread-safety in the tests where this is used, but consider `Interlocked.Increment`

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockUnresponsiveHttpClientHandler.cs [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+        /// Callback that is triggered when a request has started.
+        /// This can be used to coordinate with a task on a separate
+        /// thread that should run only after a request has started.
+        /// For example, we may want to test aborting a request after
+        /// the request has started.
+        /// </summary>
+        public Action OnRequestStarted { get; set; }
+
+        public HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
```
Comment (corranrogue9) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockUnresponsiveHttpClientHandler.cs:35]: I think this is a case where the non-async method should call the async one. I think this will do what you want:

```
protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
{
    await Task.Delay(TimeSpan.FromMilliseconds(MaxWaitTimeMillis), cancellationToken).ConfigureAwait(false);
    throw new Exception(
        $"{nameof(MockUnresponsiveHttpClientHandler)} has exceeded maximum {MaxWaitTimeMillis}ms wait time in request."
        + "Review your logic to ensure the request is cancelled in time.");
}

public HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
{
    return SendAsync(request, cancellationToken).GetAwaiter().GetResult();
}
```

File: src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+// </copyright>
+//---------------------------------------------------------------------
+
+using System.Net.Http;
+
+namespace Microsoft.OData.Client
+{
+    /// <summary>
+    /// Provides the <see cref="HttpClientHandler"/> instance(s) used by OData Client
+    /// to execute requests.
```
Comment (ElizabethOkerio) [inline src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs:13]: is the `HttpcClienthHandler` used to execute requests? can the documentation be improved to mirror what the `HttpClienthHandler` does?

Comment (habbes) [reply src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs:13]: Technically, the `HttpClientHandler` does execute the requests. The `HttpClient`' `SendAsync` method, calls the `HttpClientHandler`'s `SendAsync` method.

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockHttpClientHandlerProvider.cs (hunk 15 above)
Comment (habbes) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockHttpClientHandlerProvider.cs:23]: Yeah, this is left over from a previous implementation. Nice catch.

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs (hunk 13 above)
Comment (habbes) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs:107]: Here I wanted to make sure the request is disposed *before* creating the new client. If I create a `using` block, maybe it makes sense to close before the line that creates a new client and to do away with the explicit `request.Dispose()` and `stream.Dispose()` calls.

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs (hunk 12 above)
Comment (habbes) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs:55]: `IODataResponseMessage` is not `IDisposable` though...

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockHttpClientHandler.cs (hunk 14 above)
Comment (habbes) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/MockHttpClientHandler.cs:39]: It does compile.

File: src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs (hunk 11 above)
Comment (habbes) [reply src/Microsoft.OData.Client/IHttpClientHandlerProvider.cs:27]: I had passed the http request message at the beginning, and there was consideration to pass context with more info. But I removed it after Mike's suggestion to keep things simple for 7.x

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs [hunk 19]
```diff
@@ -18,5 +28,185 @@ public void UnwrapAggregateException()
...
+                    "GET",
+                    new Uri("http://localhost"),
+                    useDefaultCredentials: false,
+                    usePostTunneling: false,
+                    new Dictionary<string, string>(),
+                    httpClientProvider);
+
+                using (var request = new HttpClientRequestMessage(args))
+                {
+                    request.Timeout = 1;
```
[the remark this answers is not in the document]
Comment (habbes) [reply test/FunctionalTests/Microsoft.OData.Client.Tests/Serialization/HttpClientRequestMessageTests.cs:170]: Could you elaborate on what race condition would occur and how a sleep would prevent that?

**Review by ElizabethOkerio — APPROVED** [review]
